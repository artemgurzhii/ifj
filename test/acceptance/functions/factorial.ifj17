Function recursive (n As Integer) As Integer
dim answer as integer
dim decremented_n as integer

if (n == 1) then
  return 1
else
decremented_n = n - 1
answer = recursive(decremented_n)
answer = answer * n
end if

return answer
End Function


Scope
dim a as integer
a = 2
a = recursive(a)
End Scope
