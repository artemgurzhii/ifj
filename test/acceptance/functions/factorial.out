.IFJcode17
JUMP Scope
LABEL recursive
CREATEFRAME
DEFVAR TF@n
POPS TF@n
DEFVAR TF@answer
DEFVAR TF@decremented_n
JUMPIFEQ RES_IF_1 TF@n int@1
JUMP RES_ELSE_1
LABEL RES_IF_1
PUSHS int@1
RETURN
JUMP END_IF_1
LABEL RES_ELSE_1
SUB TF@decremented_n TF@n int@1
PUSHS TF@decremented_n
PUSHFRAME
CALL recursive
POPFRAME
POPS TF@answer
MUL TF@answer TF@answer TF@n
JUMP END_IF_1
LABEL END_IF_1
PUSHS TF@answer
RETURN
LABEL Scope
DEFVAR GF@a
MOVE GF@a int@5
PUSHS GF@a
CALL recursive
POPS GF@a
